<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="CHOIR">
<title>CHOIR • CHOIR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="CHOIR">
<meta property="og:description" content="CHOIR">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">CHOIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown--tutorials">
    <span class="fa fa-graduation-cap"></span>
     
    Tutorials
  </a>
  <div class="dropdown-menu" aria-labelledby="dropdown--tutorials">
    <a class="dropdown-item" href="../articles/CHOIR.html">Quick start</a>
    <a class="dropdown-item" href="../articles/atlas_scale_data.html">Atlas-scale data</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    Functions
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/corceslab/CHOIR">
    <span class="fa fa-github"></span>
     
    CHOIR on GitHub
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>CHOIR</h1>
            
      
      
      <div class="d-none name"><code>CHOIR.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>CHOIR is designed to be run on Unix-based operating systems such as
macOS and linux.</p>
<p>CHOIR installation currently requires <code>remotes</code> and
<code>BiocManager</code> for installation of GitHub and Bioconductor
packages. Run the following commands to install the various dependencies
used by CHOIR:</p>
<p>First, install remotes (for installing GitHub packages) if it isn’t
already installed:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"remotes"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span></code></pre></div>
<p>Then, install BiocManager (for installing bioconductor packages) if
it isn’t already installed:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span></code></pre></div>
<p>Then, install CHOIR:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"corceslab/CHOIR"</span>, ref<span class="op">=</span><span class="st">"main"</span>, repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span>, upgrade <span class="op">=</span> <span class="st">"never"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, install the <code>dev</code> branch for the latest
updates and features (e.g., additional compatibility, improved
efficiency, etc.):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"corceslab/CHOIR"</span>, ref<span class="op">=</span><span class="st">"dev"</span>, repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span>, upgrade <span class="op">=</span> <span class="st">"never"</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides a basic example of how to run CHOIR, a
clustering algorithm for single-cell data. CHOIR is applicable to
single-cell data of any modality, including RNA, ATAC, and proteomics.
It is also applicable to multi-modal data (see <a href="https://www.choirclustering.com/articles/CHOIR.html#using-choir-with-multi-modal-data" class="external-link">Advanced
Options</a>). Detailed parameter definitions are available under the <a href="https://www.choirclustering.com/reference/index.html" class="external-link">Functions</a>
tab.</p>
<p>CHOIR is based on the premise that if clusters contain biologically
different cell types or states, a machine learning classifier that
considers features present in cells from each cluster should be able to
distinguish the clusters with a higher level of accuracy than machine
learning classifiers trained on randomly permuted cluster labels. The
use of permutation testing approaches allows CHOIR to introduce
statistical significance thresholds into the clustering process.</p>
<p>CHOIR proceeds in two main steps. First, CHOIR generates a
hierarchical clustering tree spanning the network structure of the data
from an initial “partition” in which all cells are in the same cluster
to a partition in which all cells are demonstrably overclustered. Then,
CHOIR applies a permutation testing approach using random forest
classifiers across the nodes of the hierarchical clustering tree to
determine the appropriate extent of each branch of the clustering tree,
such that the final clusters represent statistically distinct
populations.</p>
<p>You’ll need the following packages installed to run this
tutorial:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"scRNAseq"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"scRNAseq"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CHOIR</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h4>
<p>To demonstrate how to run CHOIR on a single-cell RNA-seq dataset,
we’ll use a small dataset consisting of mouse dopaminergic neurons,
originating from an experiment by La Manno et al. (2016), and available
through the Bioconductor package <code>scRNAseq</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/LaMannoBrainData.html" class="external-link">LaMannoBrainData</a></span><span class="op">(</span><span class="st">'mouse-adult'</span><span class="op">)</span><span class="op">@</span><span class="va">assays</span><span class="op">@</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="pre-processing">Pre-processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h4>
<p>CHOIR takes as input a Seurat, SingleCellExperiment, or ArchR object
containing one or more feature matrices. Input data should already have
undergone appropriate quality control and normalization steps (unless
you are using SCTransform, see <a href="https://www.choirclustering.com/articles/CHOIR.html#input-object-types" class="external-link">Advanced
Options</a>). Note that the exact pre-processing steps used are up to
the user.</p>
<p>This tutorial uses a Seurat object; for details related to other
object types, see <a href="https://www.choirclustering.com/articles/CHOIR.html#input-object-types" class="external-link">Advanced
Options</a>. First, we will create a Seurat object using the read count
matrix. For simplicity, here we’ll just exclude cells with fewer than
100 reads and genes present in less than 5 cells.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">data_matrix</span>, </span>
<span>                             min.features <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                             min.cells <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>We will now run log normalization.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="running-choir">Running CHOIR<a class="anchor" aria-label="anchor" href="#running-choir"></a>
</h2>
<p>CHOIR proceeds in two main steps:</p>
<ol style="list-style-type: decimal">
<li>CHOIR generates a hierarchical clustering tree.</li>
<li>CHOIR prunes this tree through the iterative application of a
permutation testing approach.</li>
</ol>
<p>The two steps can be run together using the function
<code><a href="../reference/CHOIR.html">CHOIR()</a></code> or separately using functions
<code><a href="../reference/buildTree.html">buildTree()</a></code> and <code><a href="../reference/pruneTree.html">pruneTree()</a></code>.</p>
<div class="section level3">
<h3 id="quick-start">Quick start:<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h3>
<p>The <code><a href="../reference/CHOIR.html">CHOIR()</a></code> function will run all of the steps of the
CHOIR algorithm in sequence. CHOIR is highly parallelized, so efficiency
greatly improves as <code>n_cores</code> is increased.</p>
<p>The default significance level used by CHOIR is
<code>alpha = 0.05</code> with Bonferroni multiple comparison
correction. Other correction methods may be less conservative, as CHOIR
applies filters that reduce the total number of tests performed (see <a href="https://www.choirclustering.com/articles/CHOIR.html#choir-parameters" class="external-link">Advanced
Options</a>).</p>
<p>We recommend using the default value of <code>alpha = 0.05</code>
with Bonferroni multiple comparison correction. For a more conservative
approach, the <code>alpha</code> value could be decreased to 0.01 or
0.001.</p>
<p>If your data has multiple batches, we highly recommend using CHOIR’s
built-in batch correction, using the
<code>batch_correction_method</code> and <code>batch_labels</code>
parameters (see <a href="https://www.choirclustering.com/articles/CHOIR.html#choir-parameters" class="external-link">Advanced
Options</a>).</p>
<p>To use separate “count split” input matrices for building vs. pruning
the clustering tree, set <code>countsplit = TRUE</code> and provide
matrix suffixes to <code>countsplit_suffix</code> (see <a href="https://www.choirclustering.com/articles/CHOIR.html#count-splitting" class="external-link">Advanced
Options</a> for details).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CHOIR.html">CHOIR</a></span><span class="op">(</span><span class="va">object</span>, </span>
<span>                n_cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-by-step">Step-by-step:<a class="anchor" aria-label="anchor" href="#step-by-step"></a>
</h3>
<p>Alternately, we can run each of the two main steps of CHOIR
individually.</p>
<p>For users who already have a set of clusters generated by a different
tool, and wish to assess whether these clusters are under- or
over-clustered, start at Step 2 (see <a href="https://www.choirclustering.com/articles/CHOIR.html#providing-pre-generated-clusters" class="external-link">Advanced
Options</a>).</p>
<div class="section level4">
<h4 id="step-1-generate-hierarchical-clustering-tree">Step 1: Generate hierarchical clustering tree<a class="anchor" aria-label="anchor" href="#step-1-generate-hierarchical-clustering-tree"></a>
</h4>
<p>CHOIR generates a hierarchical clustering tree by computing
dimensionality reduction and nearest neighbor adjacency matrices for the
data, computing clusters across a series of sampled resolutions, then
reconciling those clustering results into a consensus clustering tree
using package <code>MRtree</code>. When using the default parameter
<code>max_clusters = "auto"</code>, CHOIR applies a series of
subclustering steps to ensure that the tree is not underclustered.
Specifically, the silhouette score is assessed at each new level of the
emerging tree. When the silhouette score is maximized, each cluster at
this level is subset, a new dimensionality reduction is generated, and
the cluster is then subclustered. Each of these subtrees continues to be
subdivided until the farthest pair of nearest neighboring clusters is
found to be overclustered using the permutation test approach.</p>
<p>The user can determine which method is used for dimensionality
reduction and batch correction, as well as supply specific parameters
for dimensionality reduction, batch correction, detecting nearest
neighbors, and modularity-based clustering, as needed. If your data has
multiple batches, we highly recommend using CHOIR’s built-in batch
correction, using the <code>batch_correction_method</code> and
<code>batch_labels</code> parameters (see <a href="https://www.choirclustering.com/articles/CHOIR.html#choir-parameters" class="external-link">Advanced
Options</a>).</p>
<p>Here, we will build the hierarchical clustering tree using default
parameter settings.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildTree.html">buildTree</a></span><span class="op">(</span><span class="va">object</span>,</span>
<span>                    n_cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:52 : (Step 1/7) Checking inputs and preparing object..</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Input data:</span></span>
<span><span class="co">#&gt;  - Object type: Seurat (v5)</span></span>
<span><span class="co">#&gt;  - # of cells: 243</span></span>
<span><span class="co">#&gt;  - # of batches: 1</span></span>
<span><span class="co">#&gt;  - # of modalities: 1</span></span>
<span><span class="co">#&gt;  - ATAC data: FALSE</span></span>
<span><span class="co">#&gt;  - Countsplitting: FALSE</span></span>
<span><span class="co">#&gt;  - Assay: RNA</span></span>
<span><span class="co">#&gt;  - Layer used to build tree: data</span></span>
<span><span class="co">#&gt;  - Layer used to prune tree: data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Proceeding with the following parameters:</span></span>
<span><span class="co">#&gt;  - Intermediate data stored under key: CHOIR</span></span>
<span><span class="co">#&gt;  - Alpha: 0.05</span></span>
<span><span class="co">#&gt;  - Multiple comparison adjustment: bonferroni</span></span>
<span><span class="co">#&gt;  - Features to train RF: var</span></span>
<span><span class="co">#&gt;  - # of excluded features: 0</span></span>
<span><span class="co">#&gt;  - # of permutations: 100</span></span>
<span><span class="co">#&gt;  - # of RF trees: 50</span></span>
<span><span class="co">#&gt;  - Use variance: TRUE</span></span>
<span><span class="co">#&gt;  - Minimum accuracy: 0.5</span></span>
<span><span class="co">#&gt;  - Minimum connections: 1</span></span>
<span><span class="co">#&gt;  - Maximum repeated errors: 20</span></span>
<span><span class="co">#&gt;  - Maximum cells sampled: Inf</span></span>
<span><span class="co">#&gt;  - Downsampling rate: 0.5</span></span>
<span><span class="co">#&gt;  - Minimum reads: &gt;0 reads</span></span>
<span><span class="co">#&gt;  - Maximum clusters: auto</span></span>
<span><span class="co">#&gt;  - Minimum cluster depth: 2000</span></span>
<span><span class="co">#&gt;  - Normalization method: none</span></span>
<span><span class="co">#&gt;  - Subtree dimensionality reductions: TRUE</span></span>
<span><span class="co">#&gt;  - Dimensionality reduction method: Default</span></span>
<span><span class="co">#&gt;  - Dimensionality reduction parameters provided: No</span></span>
<span><span class="co">#&gt;  - # of variable features: Default</span></span>
<span><span class="co">#&gt;  - Batch correction method: none</span></span>
<span><span class="co">#&gt;  - Batch correction parameters provided: No</span></span>
<span><span class="co">#&gt;  - Nearest neighbor parameters provided: </span></span>
<span><span class="co">#&gt;      - verbose: FALSE</span></span>
<span><span class="co">#&gt;  - Clustering parameters provided: </span></span>
<span><span class="co">#&gt;      - algorithm: 1</span></span>
<span><span class="co">#&gt;      - group.singletons: TRUE</span></span>
<span><span class="co">#&gt;      - verbose: FALSE</span></span>
<span><span class="co">#&gt;  - # of cores: 2</span></span>
<span><span class="co">#&gt;  - Random seed: 1</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:52 : (Step 2/7) Running initial dimensionality reduction..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:52 : Preparing matrix using 'RNA' assay and 'data' slot..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:53 : Running PCA with 2000 variable features..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:53 : (Step 3/7) Generating initial nearest neighbors graph..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:53 : (Step 4/7) Identify starting clustering resolution..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:53 : Starting resolution: 0.4</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:53 : (Step 5/7) Building root clustering tree..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:53 : Identified resolution with maximum silhouette.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                       Identified 3 clusters in root tree.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:53 : (Step 6/7) Subclustering root tree..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:54 : 5% (Subtree 1/3, 112 cells), 3 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:55 : 44% (Subtree 1/3, 112 cells), 5 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:55 : 50% (Subtree 2/3, 84 cells), 5 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:55 : 51% (Subtree 2/3, 84 cells), 5 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:55 : 79% (Subtree 2/3, 84 cells), 7 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:55 : 81% (Subtree 2/3, 84 cells), 7 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:56 : 83% (Subtree 3/3, 47 cells), 7 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:57 : 99% (Subtree 3/3, 47 cells), 11 total clusters.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:57 : (Step 7/7) Compiling full clustering tree..</span></span>
<span><span class="co">#&gt;                       Full tree has 6 levels and 8 clusters.</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-2-prune-hierarchical-clustering-tree">Step 2: Prune hierarchical clustering tree<a class="anchor" aria-label="anchor" href="#step-2-prune-hierarchical-clustering-tree"></a>
</h4>
<p>After constructing the hierarchical clustering tree, CHOIR iterates
through each node of the clustering tree using a bottom-up approach. At
each node, CHOIR computes pairwise comparisons of all child clusters by
splitting cells into training and test sets, training a balanced random
forest classifier on the gene expression data of the training set, and
then predicting the cluster assignments of the cells in the test set.
This yields a prediction accuracy score which represents the degree to
which the two clusters are distinguishable.</p>
<p>In parallel, CHOIR shuffles the cluster labels and repeats the same
process. Both comparisons are repeated using bootstrapped samples
(default = 100 iterations), resulting in a permutation test that
compares the true prediction accuracy for the clusters to the prediction
accuracy for a chance division of the cells into two random groups.</p>
<p>This permutation test yields a p-value that determines whether these
clusters are slated to merge or remain separate. The significance
threshold used can be adjusted using the <code>alpha</code> parameter.
We recommend using the default value of <code>alpha = 0.05</code> with
Bonferroni multiple comparison correction. For a more conservative
approach, the <code>alpha</code> value could be decreased to 0.01 or
0.001.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pruneTree.html">pruneTree</a></span><span class="op">(</span><span class="va">object</span>, </span>
<span>                    n_cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:57 : (Step 1/2) Checking inputs and preparing object..</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Input data:</span></span>
<span><span class="co">#&gt;  - Object type: Seurat</span></span>
<span><span class="co">#&gt;  - # of cells: 243</span></span>
<span><span class="co">#&gt;  - # of batches: 1</span></span>
<span><span class="co">#&gt;  - # of modalities: 1</span></span>
<span><span class="co">#&gt;  - # of subtrees: 4</span></span>
<span><span class="co">#&gt;  - # of levels: 6</span></span>
<span><span class="co">#&gt;  - # of starting clusters: 8</span></span>
<span><span class="co">#&gt;  - Countsplitting: FALSE</span></span>
<span><span class="co">#&gt;  - Assay: RNA</span></span>
<span><span class="co">#&gt;  - Layer used to build tree: data</span></span>
<span><span class="co">#&gt;  - Layer used to prune tree: data</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Proceeding with the following parameters:</span></span>
<span><span class="co">#&gt;  - Intermediate data stored under key: CHOIR</span></span>
<span><span class="co">#&gt;  - Alpha: 0.05</span></span>
<span><span class="co">#&gt;  - Multiple comparison adjustment: bonferroni</span></span>
<span><span class="co">#&gt;  - Features to train RF: var</span></span>
<span><span class="co">#&gt;  - # of excluded features: 0</span></span>
<span><span class="co">#&gt;  - # of permutations: 100</span></span>
<span><span class="co">#&gt;  - # of RF trees: 50</span></span>
<span><span class="co">#&gt;  - Use variance: TRUE</span></span>
<span><span class="co">#&gt;  - Minimum accuracy: 0.5</span></span>
<span><span class="co">#&gt;  - Minimum connections: 1</span></span>
<span><span class="co">#&gt;  - Maximum repeated errors: 20</span></span>
<span><span class="co">#&gt;  - Distance awareness: 2</span></span>
<span><span class="co">#&gt;  - All metrics collected: FALSE</span></span>
<span><span class="co">#&gt;  - Maximum cells sampled: Inf</span></span>
<span><span class="co">#&gt;  - Downsampling rate: 0.5</span></span>
<span><span class="co">#&gt;  - Minimum reads: &gt;0 reads</span></span>
<span><span class="co">#&gt;  - Normalization method: none</span></span>
<span><span class="co">#&gt;  - Batch correction method: none</span></span>
<span><span class="co">#&gt;  - Clustering parameters provided: </span></span>
<span><span class="co">#&gt;      - algorithm: 1</span></span>
<span><span class="co">#&gt;      - group.singletons: TRUE</span></span>
<span><span class="co">#&gt;      - verbose: FALSE</span></span>
<span><span class="co">#&gt;  - # of cores: 2</span></span>
<span><span class="co">#&gt;  - Random seed: 1</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:57 : (Step 2/2) Iterating through clustering tree..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:57 : 10% (1/6 levels) in 0 min. 8 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:57 : 20% (2/6 levels) in 0 min. 7 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:58 : 30% (2/6 levels) in 0.01 min. 6 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:59 : 40% (3/6 levels) in 0.02 min. 5 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:59 : 52% (4/6 levels) in 0.03 min. 4 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:00:59 : 61% (4/6 levels) in 0.03 min. 4 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:01 : 71% (5/6 levels) in 0.07 min. 4 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:01 : 92% (6/6 levels) in 0.07 min. 4 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:01 : Completed: all clusters compared.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:01 : Identified 4 clusters.</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="plot">Plot<a class="anchor" aria-label="anchor" href="#plot"></a>
</h3>
<p>Labels for the final clusters identified by CHOIR can be found in the
<code>meta.data</code> slot of the Seurat object. Other CHOIR outputs
are stored under the <code>misc</code> slot of the Seurat object.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">object</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;      orig.ident nCount_RNA nFeature_RNA CHOIR_parent_clusters</span></span>
<span><span class="co">#&gt; 1 SeuratProject       7826         3314                    P1</span></span>
<span><span class="co">#&gt; 2 SeuratProject       3294         1895                    P2</span></span>
<span><span class="co">#&gt; 3 SeuratProject       6776         3002                    P1</span></span>
<span><span class="co">#&gt; 4 SeuratProject       7294         3192                    P2</span></span>
<span><span class="co">#&gt; 5 SeuratProject       3747         1993                    P2</span></span>
<span><span class="co">#&gt; 6 SeuratProject      11258         4363                    P2</span></span>
<span><span class="co">#&gt;   CHOIR_clusters_0.05</span></span>
<span><span class="co">#&gt; 1                   1</span></span>
<span><span class="co">#&gt; 2                   2</span></span>
<span><span class="co">#&gt; 3                   1</span></span>
<span><span class="co">#&gt; 4                   2</span></span>
<span><span class="co">#&gt; 5                   2</span></span>
<span><span class="co">#&gt; 6                   2</span></span></code></pre></div>
<p>After clustering has finished, we can generate a UMAP dimensionality
reduction and use function <code><a href="../reference/plotCHOIR.html">plotCHOIR()</a></code> to plot a
visualization of the clusters.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run UMAP</span></span>
<span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCHOIRumap.html">runCHOIRumap</a></span><span class="op">(</span><span class="va">object</span>,</span>
<span>                       reduction <span class="op">=</span> <span class="st">"P0_reduction"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating UMAP embeddings for 1 dimensionality reductions..</span></span>
<span><span class="co">#&gt; (1/1)</span></span>
<span><span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="co">#&gt; 00:01:01 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span></span>
<span><span class="co">#&gt; Also defined by 'spam'</span></span>
<span><span class="co">#&gt; 00:01:01 Read 243 rows and found 50 numeric columns</span></span>
<span><span class="co">#&gt; 00:01:01 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span></span>
<span><span class="co">#&gt; Also defined by 'spam'</span></span>
<span><span class="co">#&gt; 00:01:01 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 00:01:01 Writing NN index file to temp file /var/folders/ty/d3bmlyq53cv_6t3v60wkfr980000gr/T//Rtmp9yEUsy/file91783c2a4605</span></span>
<span><span class="co">#&gt; 00:01:01 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 00:01:01 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 00:01:02 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 00:01:03 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">#&gt; 00:01:03 Commencing optimization for 500 epochs, with 9286 positive edges</span></span>
<span><span class="co">#&gt; 00:01:03 Optimization finished</span></span>
<span><span class="co">#&gt; Warning: No assay specified, setting assay as RNA by default.</span></span>
<span><span class="co"># Plot</span></span>
<span><span class="fu"><a href="../reference/plotCHOIR.html">plotCHOIR</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Data is of class matrix. Coercing to dgCMatrix.</span></span></code></pre></div>
<p><img src="CHOIR_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<p>This function will also overlay the prediction accuracy scores among
neighboring pairs of clusters when <code>accuracy_scores</code> is set
to <code>TRUE</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotCHOIR.html">plotCHOIR</a></span><span class="op">(</span><span class="va">object</span>,</span>
<span>          accuracy_scores <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>          plot_nearest <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Data is of class matrix. Coercing to dgCMatrix.</span></span></code></pre></div>
<p><img src="CHOIR_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="advanced-options">Advanced Options<a class="anchor" aria-label="anchor" href="#advanced-options"></a>
</h2>
<div class="section level3">
<h3 id="input-object-types">Input object types<a class="anchor" aria-label="anchor" href="#input-object-types"></a>
</h3>
<div class="section level4">
<h4 id="seurat">Seurat<a class="anchor" aria-label="anchor" href="#seurat"></a>
</h4>
<p>For Seurat objects, the default assay is used if no input is provided
for parameter <code>use_assay</code>. If no input is provided for
parameter <code>use_slot</code>, the following default slot (Seurat v4)
or layer (Seurat v5) is used:</p>
<ul>
<li>If <code>use_assay</code> is either “RNA” or “sketch”, CHOIR will
look for the <code>data</code> slot</li>
<li>If <code>use_assay</code> is either “SCT” or “integrated”, CHOIR
will look for the <code>scale.data</code> slot</li>
</ul>
<p>Some Seurat v5 objects have multiple layers (e.g., “data.1”,
“data.2”, “data.2”..) with different subsets of cells stored under the
same assay. Currently, CHOIR requires a single layer containing all
cells in an assay. For these cases, please re-organize the Seurat object
prior to running CHOIR.</p>
<p>The default dimensionality reduction method for Seurat objects is
“PCA”, except in the case of ATAC-seq data, for which it is “LSI”.</p>
<p>If you would like to use SCTransform normalization rather than log
normalization, please provide raw counts and set the parameter
<code>normalization_method = "SCTransform"</code>. Note that SCTransform
has not been thoroughly tested with CHOIR.</p>
<p>Labels for the final clusters identified by CHOIR can be found in the
<code>meta.data</code> slot of the Seurat object. Other CHOIR outputs
are stored under the <code>misc</code> slot of the Seurat object.</p>
</div>
<div class="section level4">
<h4 id="singlecellexperiment">SingleCellExperiment<a class="anchor" aria-label="anchor" href="#singlecellexperiment"></a>
</h4>
<p>For SingleCellExperiment objects, only the <code>use_assay</code>
parameter is needed. If not provided, it is set to “logcounts”. Please
ensure that the relevant assay matrix includes both row and column
names.</p>
<p>The default dimensionality reduction method for SingleCellExperiment
objects is “PCA”, except in the case of ATAC-seq data, where it is
“LSI”.</p>
<p>Labels for the final clusters identified by CHOIR can be found in the
<code>colData</code> slot of the SingleCellExperiment object. Other
CHOIR outputs are stored under <code>metadata</code>.</p>
</div>
<div class="section level4">
<h4 id="archr">ArchR<a class="anchor" aria-label="anchor" href="#archr"></a>
</h4>
<p>For ArchR objects, if no input is provided for parameter
<code>ArchR_matrix</code>, the “GeneScoreMatrix” is used. If no input
for parameter <code>ArchR_depthcol</code> is provided, “nFrags” is
used.</p>
<p>The default dimensionality reduction method for ArchR objects is
“IterativeLSI”.</p>
<p>Labels for the final clusters identified by CHOIR can be found in the
<code>cellColData</code> slot of the ArchR object. Other CHOIR outputs
are stored under <code>projectMetadata</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="choir-parameters">CHOIR parameters<a class="anchor" aria-label="anchor" href="#choir-parameters"></a>
</h3>
<div class="section level4">
<h4 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h4>
<p>For datasets with multiple batches, it is recommended to apply
Harmony batch correction through CHOIR by setting the parameter
<code>batch_correction_method = "Harmony"</code>. This not only
generates Harmony-corrected dimensionality reductions, but ensures that
random forest classifer comparisons are batch-aware. Provide the name of
the cell metadata column containing batch information to parameter
<code>batch_labels</code>.</p>
<p>Use caution in applying this method if your groups of interest (e.g.,
disease vs. control) are batch-confounded AND you expect cell types
unique to each of these groups.</p>
<p>Below is an example command to run CHOIR with batch correction:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CHOIR.html">CHOIR</a></span><span class="op">(</span><span class="va">object</span>, </span>
<span>                batch_correction_method <span class="op">=</span> <span class="st">"Harmony"</span>,</span>
<span>                batch_labels <span class="op">=</span> <span class="st">"Batch"</span><span class="op">)</span> <span class="co"># Change 'Batch' to corresponding cell metadata column for your data</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="significance-level-multiple-comparison-correction">Significance level &amp; multiple comparison correction<a class="anchor" aria-label="anchor" href="#significance-level-multiple-comparison-correction"></a>
</h4>
<p>The default significance level used by CHOIR is
<code>alpha = 0.05</code> with Bonferroni multiple comparison
correction. Other correction methods may be less conservative, as
<code>CHOIR</code> applies filters that reduce the total number of tests
performed (see below).</p>
<p>We recommend using the default value of <code>alpha = 0.05</code>
with Bonferroni multiple comparison correction. For a more conservative
approach, the <code>alpha</code> value could be decreased to 0.01 or
0.001.</p>
</div>
<div class="section level4">
<h4 id="filters">Filters<a class="anchor" aria-label="anchor" href="#filters"></a>
</h4>
<p>CHOIR uses various filters to reduce the number of necessary
permutation test comparisons:</p>
<ul>
<li>Parameter <code>min_accuracy</code> indicates the minimum accuracy
for the random forest classifier predictions, below which clusters will
be automatically merged. It defaults to 0.5, higher values will identify
a more conservative set of clusters.</li>
<li>Parameter <code>min_connections</code> indicates the minimum number
of nearest neighbors between two clusters for them to be considered
‘adjacent’. If non-zero, non-adjacent clusters will never be merged.
Defaults to 1.</li>
<li>Parameter <code>max_repeat_errors</code> is used to account for
situations in which random forest classifier prediction errors are
concentrated among a few cells that are repeatedly misassigned (often
doublets or other outliers). If non-zero, repeated errors will be
evaluated and accuracy scores will be modified if the number of repeated
errors is below the value of <code>max_repeat_errors</code>. Defaults to
20.</li>
<li>Parameter <code>distance_awareness</code> represents the distance
threshold above which a cluster will not merge with another cluster.
Specifically, this value is a multiplier applied to the distance between
a cluster and its closest distinguishable neighbor; the product sets the
threshold for subsequent comparisons. The default value sets this
threshold at a 2-fold increase in distance.</li>
</ul>
</div>
<div class="section level4">
<h4 id="downsampling">Downsampling<a class="anchor" aria-label="anchor" href="#downsampling"></a>
</h4>
<p>CHOIR uses downsampling to increase efficiency for larger datasets.
Using the default parameter setting of
<code>downsampling_rate = "auto"</code>, downsampling occurs at each
random forest classifer comparison. The downsampling rate is determined
based on the overall dataset size. To disable downsampling, set
<code>downsampling_rate = 1</code>. For datasets that require batch
correction across many batches, we recommend setting
<code>downsampling_rate = 1</code>.</p>
<p>Additional downsampling can be imposed using parameter
<code>sample_max</code>, indicating the maximum number of cells used per
cluster to train/test each random forest classifier. By default, this is
not used.</p>
</div>
</div>
<div class="section level3">
<h3 id="providing-pre-generated-clusters">Providing pre-generated clusters<a class="anchor" aria-label="anchor" href="#providing-pre-generated-clusters"></a>
</h3>
<p>Users who already have a set of clusters generated by a different
tool can use CHOIR to assess whether these clusters are under- or
over-clustered using the <code><a href="../reference/pruneTree.html">pruneTree()</a></code> function.</p>
<p>If a clustering tree has been pre-generated by the user, provided
this to parameter <code>cluster_tree</code>. If not, run the CHOIR
function <code><a href="../reference/inferTree.html">inferTree()</a></code>, which generates a simple hierarchical
clustering tree from a set of clusters based on their distances in a
provided dimensionality reduction.</p>
<p>To <code><a href="../reference/pruneTree.html">pruneTree()</a></code>, provide:</p>
<ul>
<li>
<code>object</code> The input object under which the results will be
stored.</li>
<li>
<code>cluster_tree</code> A dataframe containing the cluster IDs of
each cell across the levels of a hierarchical clustering tree.</li>
<li>
<code>input_matrix</code> A matrix containing the feature x cell
data on which to train the random forest classifiers.</li>
<li>
<code>nn_matrix</code> A matrix containing the nearest neighbor
adjacency of the cells.</li>
<li>Either reduction (a matrix of dimensionality reduction cell
embeddings) if using approximate distances OR <code>dist_matrix</code>
(a distance matrix of cell to cell distances)</li>
</ul>
</div>
<div class="section level3">
<h3 id="using-choir-with-atac-seq-data">Using CHOIR with ATAC-seq data<a class="anchor" aria-label="anchor" href="#using-choir-with-atac-seq-data"></a>
</h3>
<p>If your input data is ATAC-seq data, make sure to set input parameter
<code>atac = TRUE</code>, in order to use the correct defaults. For
example, CHOIR uses 25000 variable features by default for ATAC-seq
data, in comparison to 2000 by default for other data. If your data
exclusively contains ATAC-seq data (no other modalities), it is
recommended to set parameter <code>use_variance = FALSE</code>.</p>
<p>CHOIR is compatible with both ArchR and Signac approaches to ATAC-seq
analysis, but has been most thoroughly tested with ArchR objects.</p>
<p>Below is an example command to run CHOIR with ATAC-seq data:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ArchR_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CHOIR.html">CHOIR</a></span><span class="op">(</span><span class="va">ArchR_object</span>,</span>
<span>                      atac <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      use_variance <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="using-choir-with-multi-modal-data">Using CHOIR with multi-modal data<a class="anchor" aria-label="anchor" href="#using-choir-with-multi-modal-data"></a>
</h3>
<p>If you are providing input data which contains multiple modalities
from the same cell, such as RNA + ATAC multiome data, you must provide
multiple values (as a vector) to parameters <code>use_assay</code> and
<code>use_slot</code> (for Seurat objects), <code>use_assay</code> for
(SingleCellExperiment objects), or <code>ArchR_matrix</code> and
<code>ArchR_depthcol</code> for ArchR objects. Please make sure your
object is limited to the set of shared cells among the modalities, or
you may encounter errors.</p>
<p>If you want to use different methods of normalization, dimensionality
reduction, number of variable features, or batch correction for each
modality, input to related parameters must also be provided as vectors
or lists with one value per modality. CHOIR creates a joint
dimensionality reduction for the modalities and uses all provided
feature matrices as joint input to train/test random forest
classifiers.</p>
<p>CHOIR’s approach to distance approximation is not currently
compatible with multi-modal Seurat or SingleCellExperiment objects.
Thus, for maximal efficiency, we recommend using ArchR objects for
multi-modal analysis with CHOIR. Alternately, distance approximation can
be disabled with <code>distance_approx = FALSE</code>, but may slow down
computation.</p>
<p>Below is an example command to run CHOIR with joint RNA-seq and
ATAC-seq data:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ArchR_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CHOIR.html">CHOIR</a></span><span class="op">(</span><span class="va">ArchR_object</span>,</span>
<span>                      ArchR_matrix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GeneScoreMatrix"</span>, <span class="st">"GeneExpressionMatrix"</span><span class="op">)</span>,</span>
<span>                      ArchR_depthcol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nFrags"</span>, <span class="st">"Gex_nUMI"</span><span class="op">)</span>,</span>
<span>                      atac <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="count-splitting">Count splitting<a class="anchor" aria-label="anchor" href="#count-splitting"></a>
</h3>
<p>Count splitting is a method developed by Neufeld et al. (<a href="https://doi.org/10.1093/biostatistics/kxac047" class="external-link">paper</a>) to split
a count matrix into two matrices with the same dimensions, to avoid
“double dipping” in cross validation of clusters. When parameter
<code>countsplit = TRUE</code>, CHOIR accepts count split input
matrices. One matrix will be used to calculate highly variable features,
dimensionality reductions, nearest neighbor adjacency matrices, and the
initial clustering tree. The other matrix will be exclusively used as
input to the random forest classifiers, in order to decide when clusters
should or should not be merged.</p>
<p>We can take our Seurat object and run count splitting using the
following code, which will extract our existing count matrix, run
function <code><a href="https://rdrr.io/pkg/countsplit/man/countsplit.html" class="external-link">countsplit::countsplit</a></code>, and store the resulting
matrices back in our object with added suffixes (default is “_1” and
“_2”). Use parameter <code>normalization_method</code> to apply
normalization and store the normalized count split matrices.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCountSplit.html">runCountSplit</a></span><span class="op">(</span><span class="va">object</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:04 : Checking inputs and preparing object..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:04 : Fetching matrix 1 of 1..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:04 : Preparing matrix using 'RNA' assay and 'counts' slot..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:04 : Countsplitting matrix..</span></span>
<span><span class="co">#&gt; As no overdispersion parameters were provided, Poisson count splitting will be performed.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:04 : Storing countsplit matrices..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:04 : Storing layer 'counts_1' under assay 'RNA' in Seurat object.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:04 : Storing layer 'counts_2' under assay 'RNA' in Seurat object.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:04 : Running log normalization on countsplit matrices..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : Storing normalized countsplit matrices..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : Storing layer 'counts_log_1' under assay 'RNA' in Seurat object.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : Storing layer 'counts_log_2' under assay 'RNA' in Seurat object.</span></span></code></pre></div>
<p>CHOIR can then be run using the resulting matrices.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/CHOIR.html">CHOIR</a></span><span class="op">(</span><span class="va">object</span>, </span>
<span>                use_slot <span class="op">=</span> <span class="st">"counts_log"</span>, </span>
<span>                countsplit <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                n_cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; ----------------------------------------</span></span>
<span><span class="co">#&gt; - CHOIR - Part 1: Build clustering tree</span></span>
<span><span class="co">#&gt; ----------------------------------------</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : (Step 1/7) Checking inputs and preparing object..</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Input data:</span></span>
<span><span class="co">#&gt;  - Object type: Seurat (v5)</span></span>
<span><span class="co">#&gt;  - # of cells: 243</span></span>
<span><span class="co">#&gt;  - # of batches: 1</span></span>
<span><span class="co">#&gt;  - # of modalities: 1</span></span>
<span><span class="co">#&gt;  - ATAC data: FALSE</span></span>
<span><span class="co">#&gt;  - Countsplitting: TRUE</span></span>
<span><span class="co">#&gt;  - Assay: RNA</span></span>
<span><span class="co">#&gt;  - Layer used to build tree: counts_log_1</span></span>
<span><span class="co">#&gt;  - Layer used to prune tree: counts_log_2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Proceeding with the following parameters:</span></span>
<span><span class="co">#&gt;  - Intermediate data stored under key: CHOIR</span></span>
<span><span class="co">#&gt;  - Alpha: 0.05</span></span>
<span><span class="co">#&gt;  - Multiple comparison adjustment: bonferroni</span></span>
<span><span class="co">#&gt;  - Features to train RF: var</span></span>
<span><span class="co">#&gt;  - # of excluded features: 0</span></span>
<span><span class="co">#&gt;  - # of permutations: 100</span></span>
<span><span class="co">#&gt;  - # of RF trees: 50</span></span>
<span><span class="co">#&gt;  - Use variance: TRUE</span></span>
<span><span class="co">#&gt;  - Minimum accuracy: 0.5</span></span>
<span><span class="co">#&gt;  - Minimum connections: 1</span></span>
<span><span class="co">#&gt;  - Maximum repeated errors: 20</span></span>
<span><span class="co">#&gt;  - Maximum cells sampled: Inf</span></span>
<span><span class="co">#&gt;  - Downsampling rate: 0.5</span></span>
<span><span class="co">#&gt;  - Minimum reads: &gt;0 reads</span></span>
<span><span class="co">#&gt;  - Maximum clusters: auto</span></span>
<span><span class="co">#&gt;  - Minimum cluster depth: 2000</span></span>
<span><span class="co">#&gt;  - Normalization method: none</span></span>
<span><span class="co">#&gt;  - Subtree dimensionality reductions: TRUE</span></span>
<span><span class="co">#&gt;  - Dimensionality reduction method: Default</span></span>
<span><span class="co">#&gt;  - Dimensionality reduction parameters provided: No</span></span>
<span><span class="co">#&gt;  - # of variable features: Default</span></span>
<span><span class="co">#&gt;  - Batch correction method: none</span></span>
<span><span class="co">#&gt;  - Batch correction parameters provided: No</span></span>
<span><span class="co">#&gt;  - Nearest neighbor parameters provided: </span></span>
<span><span class="co">#&gt;      - verbose: FALSE</span></span>
<span><span class="co">#&gt;  - Clustering parameters provided: </span></span>
<span><span class="co">#&gt;      - algorithm: 1</span></span>
<span><span class="co">#&gt;      - group.singletons: TRUE</span></span>
<span><span class="co">#&gt;      - verbose: FALSE</span></span>
<span><span class="co">#&gt;  - # of cores: 2</span></span>
<span><span class="co">#&gt;  - Random seed: 1</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : (Step 2/7) Running initial dimensionality reduction..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : Preparing matrix using 'RNA' assay and 'counts_log_1' slot..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : Running PCA with 2000 variable features..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : (Step 3/7) Generating initial nearest neighbors graph..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : (Step 4/7) Identify starting clustering resolution..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : Starting resolution: 0.4</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : (Step 5/7) Building root clustering tree..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : Identified resolution with maximum silhouette.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;                       Identified 3 clusters in root tree.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:05 : (Step 6/7) Subclustering root tree..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:06 : 6% (Subtree 1/3, 136 cells), 3 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:06 : 11% (Subtree 1/3, 136 cells), 4 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:09 : 53% (Subtree 1/3, 136 cells), 7 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:10 : 59% (Subtree 2/3, 83 cells), 7 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:10 : 61% (Subtree 2/3, 83 cells), 7 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:10 : 88% (Subtree 2/3, 83 cells), 9 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:10 : 90% (Subtree 2/3, 83 cells), 9 total clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:10 : 91% (Subtree 3/3, 24 cells), 9 total clusters.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:10 : (Step 7/7) Compiling full clustering tree..</span></span>
<span><span class="co">#&gt;                       Full tree has 6 levels and 7 clusters.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ----------------------------------------</span></span>
<span><span class="co">#&gt; - CHOIR - Part 2: Prune clustering tree</span></span>
<span><span class="co">#&gt; ----------------------------------------</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:10 : (Step 1/2) Checking inputs and preparing object..</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Input data:</span></span>
<span><span class="co">#&gt;  - Object type: Seurat</span></span>
<span><span class="co">#&gt;  - # of cells: 243</span></span>
<span><span class="co">#&gt;  - # of batches: 1</span></span>
<span><span class="co">#&gt;  - # of modalities: 1</span></span>
<span><span class="co">#&gt;  - # of subtrees: 4</span></span>
<span><span class="co">#&gt;  - # of levels: 6</span></span>
<span><span class="co">#&gt;  - # of starting clusters: 7</span></span>
<span><span class="co">#&gt;  - Countsplitting: TRUE</span></span>
<span><span class="co">#&gt;  - Assay: RNA</span></span>
<span><span class="co">#&gt;  - Layer used to build tree: counts_log_1</span></span>
<span><span class="co">#&gt;  - Layer used to prune tree: counts_log_2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Proceeding with the following parameters:</span></span>
<span><span class="co">#&gt;  - Intermediate data stored under key: CHOIR</span></span>
<span><span class="co">#&gt;  - Alpha: 0.05</span></span>
<span><span class="co">#&gt;  - Multiple comparison adjustment: bonferroni</span></span>
<span><span class="co">#&gt;  - Features to train RF: var</span></span>
<span><span class="co">#&gt;  - # of excluded features: 0</span></span>
<span><span class="co">#&gt;  - # of permutations: 100</span></span>
<span><span class="co">#&gt;  - # of RF trees: 50</span></span>
<span><span class="co">#&gt;  - Use variance: TRUE</span></span>
<span><span class="co">#&gt;  - Minimum accuracy: 0.5</span></span>
<span><span class="co">#&gt;  - Minimum connections: 1</span></span>
<span><span class="co">#&gt;  - Maximum repeated errors: 20</span></span>
<span><span class="co">#&gt;  - Distance awareness: 2</span></span>
<span><span class="co">#&gt;  - All metrics collected: FALSE</span></span>
<span><span class="co">#&gt;  - Maximum cells sampled: Inf</span></span>
<span><span class="co">#&gt;  - Downsampling rate: 0.5</span></span>
<span><span class="co">#&gt;  - Minimum reads: &gt;0 reads</span></span>
<span><span class="co">#&gt;  - Normalization method: none</span></span>
<span><span class="co">#&gt;  - Batch correction method: none</span></span>
<span><span class="co">#&gt;  - Clustering parameters provided: </span></span>
<span><span class="co">#&gt;      - algorithm: 1</span></span>
<span><span class="co">#&gt;      - group.singletons: TRUE</span></span>
<span><span class="co">#&gt;      - verbose: FALSE</span></span>
<span><span class="co">#&gt;  - # of cores: 2</span></span>
<span><span class="co">#&gt;  - Random seed: 1</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:10 : (Step 2/2) Iterating through clustering tree..</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:10 : 12% (1/6 levels) in 0 min. 6 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:10 : 21% (2/6 levels) in 0 min. 6 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:11 : 30% (2/6 levels) in 0.01 min. 6 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:12 : 40% (3/6 levels) in 0.03 min. 5 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:13 : 51% (4/6 levels) in 0.04 min. 4 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:13 : 61% (4/6 levels) in 0.04 min. 4 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:14 : 71% (5/6 levels) in 0.07 min. 4 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:15 : 92% (6/6 levels) in 0.07 min. 4 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:15 : Checking for underclustering in 2 clusters.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:15 : Additional comparisons necessary. 4 clusters remaining.</span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:15 : Completed: all clusters compared.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2025-04-10 00:01:15 : Identified 4 clusters.</span></span></code></pre></div>
<p>Alternately, users can provide their own count split matrices. They
must share a prefix, which is provided to <code>use_slot</code> for
Seurat objects or <code>use_assay</code> for SingleCellExperiment
objects. The unique suffixes should be provided to
<code>countsplit_suffix</code> as a character vector.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Cathrine Sant.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
