<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="CHOIR">
<title>CHOIR • CHOIR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="CHOIR">
<meta property="og:description" content="CHOIR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">CHOIR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../articles/CHOIR.html">
    <span class="fa fa-graduation-cap"></span>
     
    Quick Tutorial
  </a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    Functions
  </a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/corceslab/CHOIR">
    <span class="fa fa-github"></span>
     
    CHOIR on GitHub
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>CHOIR</h1>
            
      
      
      <div class="d-none name"><code>CHOIR.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the package, please use the following:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"corceslab/CHOIR"</span>, ref<span class="op">=</span><span class="st">"main"</span>, repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span>, upgrade <span class="op">=</span> <span class="st">"never"</span><span class="op">)</span></code></pre></div>
<hr>
</div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This vignette provides a basic example of how to run CHOIR, a
clustering algorithm for single-cell sequencing data. CHOIR is
applicable to single-cell sequencing data of any modality, including
RNA, ATAC, and proteomics. It is also applicable to multi-modal data
(see <a href="https://www.choirclustering.com/articles/CHOIR.html#advanced-options" class="external-link">Advanced
Options</a>).</p>
<p>CHOIR is based on the premise that if clusters contain biologically
different cell types or states, a machine learning classifier that
considers features present in cells from each cluster should be able to
distinguish the clusters with a higher level of accuracy than machine
learning classifiers trained on randomly permuted cluster labels. The
use of permutation testing approaches allows CHOIR to introduce
statistical significance thresholds into the clustering process.</p>
<p>CHOIR proceeds in two main steps. First, CHOIR generates a
hierarchical clustering tree spanning the network structure of the data
from an initial “partition” in which all cells are in the same cluster
to a partition in which all cells are demonstrably overclustered. Then,
CHOIR applies a permutation testing approach using random forest
classifiers across the nodes of the hierarchical clustering tree to
determine the appropriate extent of each branch of the clustering tree,
such that the final clusters represent statistically distinct
populations.</p>
<p>You’ll need the following packages installed to run this
tutorial:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CHOIR</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></code></pre></div>
<div class="section level4">
<h4 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h4>
<p>To demonstrate how to run CHOIR on a single-cell RNA-seq dataset,
we’ll use a small dataset consisting of mouse dopaminergic neurons,
originating from an experiment by La Manno et al. (2016), and available
through the Bioconductor package <code>scRNAseq</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">data_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/LaMannoBrainData.html" class="external-link">LaMannoBrainData</a></span><span class="op">(</span><span class="st">'mouse-adult'</span><span class="op">)</span><span class="op">@</span><span class="va">assays</span><span class="op">@</span><span class="va">data</span><span class="op">$</span><span class="va">counts</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">data_matrix</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="pre-processing">Pre-processing<a class="anchor" aria-label="anchor" href="#pre-processing"></a>
</h4>
<p>CHOIR takes as input a Seurat, SingleCellExperiment, or ArchR object
containing one or more feature matrices. Input data should already have
undergone appropriate quality control and normalization steps (unless
you are using SCTransform, see <a href="https://www.choirclustering.com/articles/CHOIR.html#advanced-options" class="external-link">Advanced
Options</a>). Note that the exact pre-processing steps used are up to
the user.</p>
<p>This tutorial uses a Seurat object; for details related to other
object types, see <a href="https://www.choirclustering.com/articles/CHOIR.html#advanced-options" class="external-link">Advanced
Options</a>. First, we will create a Seurat object using the read count
matrix. For simplicity, here we’ll just exclude cells with fewer than
100 reads and genes present in less than 5 cells.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span><span class="va">data_matrix</span>, 
                                    min.features <span class="op">=</span> <span class="fl">100</span>,
                                    min.cells <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></code></pre></div>
<p>We will now run log normalization.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span></code></pre></div>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="running-choir">Running CHOIR<a class="anchor" aria-label="anchor" href="#running-choir"></a>
</h2>
<p>CHOIR proceeds in two main steps:</p>
<ol style="list-style-type: decimal">
<li>CHOIR generates a hierarchical clustering tree.</li>
<li>CHOIR prunes this tree through the iterative application of a
permutation testing approach.</li>
</ol>
<p>The two steps can be run together using the function
<code><a href="https://rdrr.io/pkg/CHOIR/man/CHOIR-package.html" class="external-link">CHOIR()</a></code> or separately using functions
<code><a href="../reference/buildTree.html">buildTree()</a></code> and <code><a href="../reference/pruneTree.html">pruneTree()</a></code>.</p>
<div class="section level3">
<h3 id="quick-start">Quick start:<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h3>
<p>The <code><a href="https://rdrr.io/pkg/CHOIR/man/CHOIR-package.html" class="external-link">CHOIR()</a></code> function will run all of the steps of the
CHOIR algorithm in sequence. CHOIR is highly parallelized, so efficiency
greatly improves as <code>n_cores</code> is increased.</p>
<p>The default significance level used by CHOIR is <span class="math inline">\(\alpha = 0.05\)</span> with Bonferroni multiple
comparison correction. Other correction methods may be less
conservative, as CHOIR applies filters that reduce the total number of
tests performed (see <a href="https://www.choirclustering.com/articles/CHOIR.html#advanced-options" class="external-link">Advanced
Options</a>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CHOIR/man/CHOIR-package.html" class="external-link">CHOIR</a></span><span class="op">(</span><span class="va">seurat_object</span>, 
                       n_cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="step-by-step">Step-by-step:<a class="anchor" aria-label="anchor" href="#step-by-step"></a>
</h3>
<p>Alternately, we can run each of the two main steps of CHOIR
individually.</p>
<p>For users who already have a set of clusters generated by a different
tool, and wish to assess whether these clusters are under- or
over-clustered, start at Step 2 (see <a href="https://www.choirclustering.com/articles/CHOIR.html#advanced-options" class="external-link">Advanced
Options</a>).</p>
<div class="section level4">
<h4 id="step-1-generate-hierarchical-clustering-tree">Step 1: Generate hierarchical clustering tree<a class="anchor" aria-label="anchor" href="#step-1-generate-hierarchical-clustering-tree"></a>
</h4>
<p>CHOIR generates a hierarchical clustering tree by computing
dimensionality reduction and nearest neighbor adjacency matrices for the
data, computing clusters across a series of sampled resolutions, then
reconciling those clustering results into a consensus clustering tree
using package <code>MRtree</code>. When using the default parameter
<code>max_clusters = 'auto'</code>, CHOIR applies a series of
subclustering steps to ensure that the tree is not underclustered.
Specifically, the silhouette score is assessed at each new level of the
emerging tree. When the silhouette score is maximized, each cluster at
this level is subset, a new dimensionality reduction is generated, and
the cluster is then subclustered. Each of these subtrees continues to be
subdivided until the farthest pair of nearest neighboring clusters is
found to be overclustered using the permutation test approach.</p>
<p>The user can determine which method is used for dimensionality
reduction and batch correction, as well as supply specific parameters
for dimensionality reduction, batch correction, detecting nearest
neighbors, and modularity-based clustering, as needed.</p>
<p>Here, we will build the hierarchical clustering tree using default
parameter settings.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/buildTree.html">buildTree</a></span><span class="op">(</span><span class="va">seurat_object</span>,
                           n_cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Input data:</span>
<span class="co">#&gt;  - Object type: Seurat (v5)</span>
<span class="co">#&gt;  - # of cells: 243</span>
<span class="co">#&gt;  - # of modalities: 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Proceeding with the following parameters:</span>
<span class="co">#&gt;  - Intermediate data stored under key: CHOIR</span>
<span class="co">#&gt;  - Normalization method: none</span>
<span class="co">#&gt;  - Subtree dimensionality reductions: TRUE</span>
<span class="co">#&gt;  - Dimensionality reduction method: Default</span>
<span class="co">#&gt;  - # of variable features: Default</span>
<span class="co">#&gt;  - Batch correction method: none</span>
<span class="co">#&gt;  - Maximum clusters: auto</span>
<span class="co">#&gt;  - Minimum cluster depth: 2000</span>
<span class="co">#&gt;  - Distance approximation: TRUE</span>
<span class="co">#&gt;  - Alpha: 0.05</span>
<span class="co">#&gt;  - Multiple comparison adjustment: bonferroni</span>
<span class="co">#&gt;  - Features to train RF: var</span>
<span class="co">#&gt;  - # of excluded features: 0</span>
<span class="co">#&gt;  - # of permutations: 100</span>
<span class="co">#&gt;  - # of RF trees: 50</span>
<span class="co">#&gt;  - Use variance: TRUE</span>
<span class="co">#&gt;  - Minimum accuracy: 0.5</span>
<span class="co">#&gt;  - Minimum connections: 1</span>
<span class="co">#&gt;  - Maximum repeated errors: 20</span>
<span class="co">#&gt;  - Maximum cells sampled: Inf</span>
<span class="co">#&gt;  - Downsampling rate: 1</span>
<span class="co">#&gt;  - # of cores: 2</span>
<span class="co">#&gt;  - Random seed: 1</span>
<span class="co">#&gt; 2023-10-20 13:08:46 : (Step 1/6) Running initial dimensionality reduction..</span>
<span class="co">#&gt;                       Preparing matrix using 'RNA' assay and 'data' slot..</span>
<span class="co">#&gt;                       Running PCA with 2000 variable features..</span>
<span class="co">#&gt; 2023-10-20 13:08:47 : (Step 2/6) Generating initial nearest neighbors graph..</span>
<span class="co">#&gt; 2023-10-20 13:08:47 : (Step 3/6) Identify starting clustering resolution..</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;                       Starting resolution: 0.4</span>
<span class="co">#&gt; 2023-10-20 13:08:47 : (Step 4/6) Building parent clustering tree..</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;                       Identified 3 clusters in parent tree.</span>
<span class="co">#&gt; 2023-10-20 13:08:48 : (Step 5/6) Subclustering parent tree..</span>
<span class="co">#&gt; 2023-10-20 13:08:49 : 5% (Subtree 1/3, 112 cells), 3 total clusters.</span>
<span class="co">#&gt; 2023-10-20 13:08:57 : 44% (Subtree 1/3, 112 cells), 7 total clusters.</span>
<span class="co">#&gt; 2023-10-20 13:08:58 : 50% (Subtree 2/3, 84 cells), 7 total clusters.</span>
<span class="co">#&gt; 2023-10-20 13:08:58 : 51% (Subtree 2/3, 84 cells), 7 total clusters.</span>
<span class="co">#&gt; 2023-10-20 13:08:59 : 79% (Subtree 2/3, 84 cells), 9 total clusters.</span>
<span class="co">#&gt; 2023-10-20 13:08:59 : 81% (Subtree 2/3, 84 cells), 9 total clusters.</span>
<span class="co">#&gt; 2023-10-20 13:09:00 : 83% (Subtree 3/3, 47 cells), 9 total clusters.</span>
<span class="co">#&gt; 2023-10-20 13:09:04 : 99% (Subtree 3/3, 47 cells), 13 total clusters.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2023-10-20 13:09:04 : (Step 6/6) Compiling full clustering tree..</span>
<span class="co">#&gt;                       Full tree has 6 levels and 10 clusters.</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="step-2-prune-hierarchical-clustering-tree">Step 2: Prune hierarchical clustering tree<a class="anchor" aria-label="anchor" href="#step-2-prune-hierarchical-clustering-tree"></a>
</h4>
<p>After constructing the hierarchical clustering tree, CHOIR iterates
through each node of the clustering tree using a bottom-up approach. At
each node, CHOIR computes pairwise comparisons of all child clusters by
splitting cells into training and test sets, training a balanced random
forest classifier on the gene expression data of the training set, and
then predicting the cluster assignments of the cells in the test set.
This yields a prediction accuracy score which represents the degree to
which the two clusters are distinguishable.</p>
<p>In parallel, CHOIR shuffles the cluster labels and repeats the same
process. Both comparisons are repeated using bootstrapped samples
(default = 100 iterations), resulting in a permutation test that
compares the true prediction accuracy for the clusters to the prediction
accuracy for a chance division of the cells into two random groups.</p>
<p>This permutation test yields a p-value that determines whether these
clusters are slated to merge or remain separate. The significance
threshold used can be adjusted using the <code>alpha</code>
parameter.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pruneTree.html">pruneTree</a></span><span class="op">(</span><span class="va">seurat_object</span>, 
                           n_cores <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; 2023-10-20 13:09:04 : (Step 1/2) Preparing object..</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Input data:</span>
<span class="co">#&gt;  - Object type: Seurat</span>
<span class="co">#&gt;  - # of cells: 243</span>
<span class="co">#&gt;  - # of modalities: 1</span>
<span class="co">#&gt;  - # of subtrees: 4</span>
<span class="co">#&gt;  - # of levels: 6</span>
<span class="co">#&gt;  - # of starting clusters: 10</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Proceeding with the following parameters:</span>
<span class="co">#&gt;  - Intermediate data stored under key: CHOIR</span>
<span class="co">#&gt;  - Alpha: 0.05</span>
<span class="co">#&gt;  - Multiple comparison adjustment: bonferroni</span>
<span class="co">#&gt;  - Features to train RF: var</span>
<span class="co">#&gt;  - # of excluded features: 0</span>
<span class="co">#&gt;  - # of permutations: 100</span>
<span class="co">#&gt;  - # of RF trees: 50</span>
<span class="co">#&gt;  - Use variance: TRUE</span>
<span class="co">#&gt;  - Minimum accuracy: 0.5</span>
<span class="co">#&gt;  - Minimum connections: 1</span>
<span class="co">#&gt;  - Maximum repeated errors: 20</span>
<span class="co">#&gt;  - Distance awareness: 2</span>
<span class="co">#&gt;  - Distance approximation: TRUE</span>
<span class="co">#&gt;  - Maximum cells sampled: Inf</span>
<span class="co">#&gt;  - Downsampling rate: 1</span>
<span class="co">#&gt;  - # of cores: 2</span>
<span class="co">#&gt;  - Random seed: 1</span>
<span class="co">#&gt; 2023-10-20 13:09:05 : (Step 2/2) Iterating through clustering tree..</span>
<span class="co">#&gt; 2023-10-20 13:09:08 : 11% (2/6 levels) in 0.06 min. 8 clusters remaining.</span>
<span class="co">#&gt; 2023-10-20 13:09:09 : 20% (2/6 levels) in 0.07 min. 6 clusters remaining.</span>
<span class="co">#&gt; 2023-10-20 13:09:13 : 32% (3/6 levels) in 0.15 min. 6 clusters remaining.</span>
<span class="co">#&gt; 2023-10-20 13:09:15 : 40% (3/6 levels) in 0.17 min. 6 clusters remaining.</span>
<span class="co">#&gt; 2023-10-20 13:09:15 : 50% (4/6 levels) in 0.17 min. 6 clusters remaining.</span>
<span class="co">#&gt; 2023-10-20 13:09:19 : 60% (5/6 levels) in 0.24 min. 6 clusters remaining.</span>
<span class="co">#&gt; 2023-10-20 13:09:35 : 70% (5/6 levels) in 0.51 min. 6 clusters remaining.</span>
<span class="co">#&gt; 2023-10-20 13:09:38 : 80% (6/6 levels) in 0.56 min. 5 clusters remaining.</span>
<span class="co">#&gt; 2023-10-20 13:09:46 : 91% (6/6 levels) in 0.7 min. 5 clusters remaining.</span>
<span class="co">#&gt; 2023-10-20 13:09:47 : Checking for underclustering in 1 clusters.</span>
<span class="co">#&gt; 2023-10-20 13:09:50 : Additional comparisons necessary. 5 clusters remaining.</span>
<span class="co">#&gt; 2023-10-20 13:09:50 : Completed: all clusters compared.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  - In 2 comparisons, clusters were split due to the minimum number of nearest neighbor connections.</span>
<span class="co">#&gt;  - In 1 comparisons, clusters were merged after correction for multiple comparisons.</span>
<span class="co">#&gt;  - Final adjusted significance threshold = 0.00227.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2023-10-20 13:09:50 : Identified 5 clusters.</span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="plot">Plot<a class="anchor" aria-label="anchor" href="#plot"></a>
</h3>
<p>The final clusters identified by CHOIR can be found in the
<code>meta.data</code> slot of the Seurat object. Other CHOIR outputs
are stored under the <code>misc</code> slot of the Seurat object.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span>
<span class="co">#&gt;      orig.ident nCount_RNA nFeature_RNA CHOIR_clusters_0.05</span>
<span class="co">#&gt; 1 SeuratProject       7826         3314                   3</span>
<span class="co">#&gt; 2 SeuratProject       3294         1895                   1</span>
<span class="co">#&gt; 3 SeuratProject       6776         3002                   2</span>
<span class="co">#&gt; 4 SeuratProject       7294         3192                   1</span>
<span class="co">#&gt; 5 SeuratProject       3747         1993                   1</span>
<span class="co">#&gt; 6 SeuratProject      11258         4363                   1</span></code></pre></div>
<p>After clustering has finished, we can generate a UMAP dimensionality
reduction and use function <code><a href="../reference/plotCHOIR.html">plotCHOIR()</a></code> to plot a
visualization of the clusters.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Run UMAP</span>
<span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/runCHOIRumap.html">runCHOIRumap</a></span><span class="op">(</span><span class="va">seurat_object</span>,
                              reduction <span class="op">=</span> <span class="st">"P0_reduction"</span><span class="op">)</span>
<span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span>
<span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span>
<span class="co">#&gt; This message will be shown once per session</span>
<span class="co">#&gt; 13:09:51 UMAP embedding parameters a = 0.9922 b = 1.112</span>
<span class="co">#&gt; Found more than one class "dist" in cache; using the first, from namespace 'spam'</span>
<span class="co">#&gt; Also defined by 'BiocGenerics'</span>
<span class="co">#&gt; 13:09:51 Read 243 rows and found 50 numeric columns</span>
<span class="co">#&gt; 13:09:51 Using Annoy for neighbor search, n_neighbors = 30</span>
<span class="co">#&gt; Found more than one class "dist" in cache; using the first, from namespace 'spam'</span>
<span class="co">#&gt; Also defined by 'BiocGenerics'</span>
<span class="co">#&gt; 13:09:51 Building Annoy index with metric = cosine, n_trees = 50</span>
<span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span>
<span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span>
<span class="co">#&gt; **************************************************|</span>
<span class="co">#&gt; 13:09:51 Writing NN index file to temp file /var/folders/ty/d3bmlyq53cv_6t3v60wkfr980000gq/T//RtmpABnGgH/file96191d2a13c5</span>
<span class="co">#&gt; 13:09:51 Searching Annoy index using 1 thread, search_k = 3000</span>
<span class="co">#&gt; 13:09:51 Annoy recall = 100%</span>
<span class="co">#&gt; 13:09:52 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span>
<span class="co">#&gt; 13:09:53 Initializing from normalized Laplacian + noise (using RSpectra)</span>
<span class="co">#&gt; 13:09:54 Commencing optimization for 500 epochs, with 9286 positive edges</span>
<span class="co">#&gt; 13:09:55 Optimization finished</span>
<span class="co">#&gt; Warning: No assay specified, setting assay as RNA by default.</span>
<span class="co"># Plot</span>
<span class="fu"><a href="../reference/plotCHOIR.html">plotCHOIR</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span>
<span class="co">#&gt; Counts matrix provided is not sparse. Creating V5 assay in Seurat Object.</span></code></pre></div>
<p><img src="CHOIR_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>This function will also overlay the prediction accuracy scores among
neighboring pairs of clusters when <code>accuracy_scores</code> is set
to <code>TRUE</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotCHOIR.html">plotCHOIR</a></span><span class="op">(</span><span class="va">seurat_object</span>,
          accuracy_scores <span class="op">=</span> <span class="cn">TRUE</span>,
          plot_nearest <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Counts matrix provided is not sparse. Creating V5 assay in Seurat Object.</span></code></pre></div>
<p><img src="CHOIR_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="advanced-options">Advanced Options<a class="anchor" aria-label="anchor" href="#advanced-options"></a>
</h2>
<div class="section level3">
<h3 id="input-object-types">Input object types<a class="anchor" aria-label="anchor" href="#input-object-types"></a>
</h3>
<div class="section level4">
<h4 id="seurat">Seurat<a class="anchor" aria-label="anchor" href="#seurat"></a>
</h4>
<p>For Seurat objects, the default assay is used if no input is provided
for parameter <code>use_assay</code>. If no input is provided for
parameter <code>use_slot</code>, the following default slot (Seurat v4)
or layer (Seurat v5) is used:</p>
<ul>
<li>If <code>use_assay</code> is either “RNA” or “sketch”, CHOIR will
look for the <code>data</code> slot</li>
<li>If <code>use_assay</code> is either “SCT” or “integrated”, CHOIR
will look for the <code>scale.data</code> slot</li>
</ul>
<p>The default dimensionality reduction method for Seurat objects is
‘PCA’, except in the case of ATAC-seq data, where it is ‘LSI’.</p>
<p>If you would like to use SCTransform normalization rather than log
normalization, please provide raw counts and set the parameter
<code>normalization_method</code> to ‘SCTransform’. Note that
SCTransform has not been thoroughly tested with CHOIR.</p>
</div>
<div class="section level4">
<h4 id="singlecellexperiment">SingleCellExperiment<a class="anchor" aria-label="anchor" href="#singlecellexperiment"></a>
</h4>
<p>For SingleCellExperiment objects, only the <code>use_assay</code>
parameter is needed. If not provided, it is set to ‘logcounts’.</p>
<p>The default dimensionality reduction method for Seurat objects is
‘PCA’, except in the case of ATAC-seq data, where it is ‘LSI’.</p>
</div>
<div class="section level4">
<h4 id="archr">ArchR<a class="anchor" aria-label="anchor" href="#archr"></a>
</h4>
<p>For ArchR objects, if no input is provided for parameter
<code>ArchR_matrix</code>, the “TileMatrix” is used. If no input for
parameter <code>ArchR_depthcol</code> is provided, “nFrags” is used.</p>
<p>The default dimensionality reduction method for ArchR objects is
‘IterativeLSI’.</p>
</div>
</div>
<div class="section level3">
<h3 id="choir-parameters">CHOIR parameters<a class="anchor" aria-label="anchor" href="#choir-parameters"></a>
</h3>
<div class="section level4">
<h4 id="batch-correction">Batch correction<a class="anchor" aria-label="anchor" href="#batch-correction"></a>
</h4>
<p>For datasets with multiple batches, it is recommended to apply
Harmony batch correction through CHOIR by setting the parameter
<code>batch_correction_method</code> to ‘Harmony’. This not only
generates Harmony-corrected dimnesionality reductions, but ensures that
random forest classifer comparisons are batch-aware.</p>
<p>Use caution in applying this method if your groups of interest (e.g.,
disease vs. control) are batch-confounded AND you expect cell types
unique to each of these groups.</p>
</div>
<div class="section level4">
<h4 id="significance-level-multiple-comparison-correction">Significance level &amp; multiple comparison correction<a class="anchor" aria-label="anchor" href="#significance-level-multiple-comparison-correction"></a>
</h4>
<p>The default significance level used by CHOIR is <span class="math inline">\(\alpha = 0.05\)</span> with Bonferroni multiple
comparison correction. Other correction methods may be less
conservative, as <code>CHOIR</code> applies filters that reduce the
total number of tests performed (see below).</p>
</div>
<div class="section level4">
<h4 id="filters">Filters<a class="anchor" aria-label="anchor" href="#filters"></a>
</h4>
<p>CHOIR uses various filters to reduce the number of necessary
permutation test comparisons:</p>
<ul>
<li>Parameter <code>min_accuracy</code> indicates the minimum accuracy
for the random forest classifier predictions, below which clusters will
be automatically merged. It defaults to 0.5, higher values will identify
a more conservative set of clusters.</li>
<li>Parameter <code>min_connections</code> indicates the minimum number
of nearest neighbors between two clusters for them to be considered
‘adjacent’. If non-zero, non-adjacent clusters will never be merged.
Defaults to 1.</li>
<li>Parameter <code>max_repeat_errors</code> is used to account for
situations in which random forest classifier prediction errors are
concentrated among a few cells that are repeatedly misassigned (often
doublets or other outliers). If non-zero, repeated errors will be
evaluated and accuracy scores will be modified if the number of repeated
errors is below the value of <code>max_repeat_errors</code>. Defaults to
20.</li>
<li>Parameter <code>distance_awareness</code> represents the distance
threshold above which a cluster will not merge with another cluster.
Specifically, this value is a multiplier applied to the distance between
a cluster and its closest distinguishable neighbor; the product sets the
threshold for subsequent comparisons. The default value sets this
threshold at a 2-fold increase in distance.</li>
</ul>
</div>
<div class="section level4">
<h4 id="downsampling">Downsampling<a class="anchor" aria-label="anchor" href="#downsampling"></a>
</h4>
<p>CHOIR uses downsampling to increase efficiency for larger datasets.
Datasets above 5000 cells are automatically downsampled according to
their size. Downsampling occurs at each random forest classifer
comparison, using the default parameter setting
<code>downsampling_rate = "auto"</code>.</p>
<p>Additional downsampling can be imposed using parameter
<code>sample_max</code>, indicating the maximum number of cells used per
cluster to train/test each random forest classifier. The default value
does not cap the number of cells used.</p>
</div>
</div>
<div class="section level3">
<h3 id="providing-pre-generated-clusters">Providing pre-generated clusters<a class="anchor" aria-label="anchor" href="#providing-pre-generated-clusters"></a>
</h3>
<p>For users who already have a set of clusters generated by a different
tool, and wish to assess whether these clusters are under- or
over-clustered, only need the <code><a href="../reference/pruneTree.html">pruneTree()</a></code> function.</p>
<p>To <code><a href="../reference/pruneTree.html">pruneTree()</a></code>, provide:</p>
<ul>
<li>
<code>object</code> The input object under which the results will be
stored.</li>
<li>
<code>cluster_tree</code> A dataframe containing the cluster IDs of
each cell across the levels of a hierarchical clustering tree. This can
be generated from a single level of clusters using function
<code>createHierachy()</code> (IN DEVELOPMENT).</li>
<li>
<code>input_matrix</code> A matrix containing the feature x cell
data on which to train the random forest classifiers.</li>
<li>
<code>nn_matrix</code> A matrix containing the nearest neighbor
adjacency of the cells.</li>
<li>Either reduction (a matrix of dimensionality reduction cell
embeddings) if using approximate distances OR <code>dist_matrix</code>
(a distance matrix of cell to cell distances)</li>
</ul>
</div>
<div class="section level3">
<h3 id="using-choir-with-atac-seq-data">Using CHOIR with ATAC-seq data<a class="anchor" aria-label="anchor" href="#using-choir-with-atac-seq-data"></a>
</h3>
<p>If your input data is ATAC-seq data, make sure to set input parameter
<code>atac</code> to <code>TRUE</code>, in order to use the correct
defaults. For example, CHOIR uses 25000 variable features by default for
ATAC-seq data, in comparison to 2000 by default for other data. If your
data exclusively contains ATAC-seq data (no other modalities), it is
recommended to set parameter <code>use_variance</code> to
<code>FALSE</code>.</p>
</div>
<div class="section level3">
<h3 id="using-choir-with-multi-modal-data">Using CHOIR with multi-modal data<a class="anchor" aria-label="anchor" href="#using-choir-with-multi-modal-data"></a>
</h3>
<p>If you are providing input data which contains multiple modalities
from the same cell, such as RNA + ATAC multiome data, you must provide
multiple values (as a vector) to parameters <code>use_assay</code> and
<code>use_slot</code> (for Seurat objects), <code>use_assay</code> for
(SingleCellExperiment objects), or <code>ArchR_matrix</code>and
<code>ArchR_depthcol</code> for ArchR objects.</p>
<p>If you want to use different methods of normalization, dimensionality
reduction, number of variable features, or batch correction for each
modality, input to related parameters must also be provided as vectors
or lists with one value per modality.</p>
<p>CHOIR creates a joint dimensionality reduction for the modalities and
uses all provided feature matrices as joint input to train/test random
forest classifiers.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Cathrine Petersen.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.4.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
